app/controllers/collections_controller.rb

```diff
@@ -87,4 +117,62 @@ def set_collection
   def collection_params
     params.require(:collection).permit(:title, :description, :image, :image_url, :public, {:keywords => []}, {:material_ids => []}, {:event_ids => []})
   end
+
+  # Filter collection items based on a type
+  def item_class
+    case params[:type]
+    when 'Event'
+      return Event if TeSS::Config.feature['events']
+    when 'Material'
+      return Material if TeSS::Config.feature['materials']
+    end
+
+    raise ActiveRecord::AccessDenied
+  end
+
+  # since we have not checked all items we have to do a little work
+  # to add and remove only those that were checked now, with bulk queries
+  # for good performance.
+  def update_collection_items!
+    # remove unselected ones, if any exist
+    CollectionItem.where(collection_id: @collection.id,
+                         resource_id: removed_collection_item_ids,
+                         resource_type: item_class.name).delete_all # no callbacks!
+
+    # bulk insert
+    max_order = @collection.items.maximum(:order) || 0
+    # https://www.bigbinary.com/blog/bulk-insert-support-in-rails-6#1-performing-bulk-inserts-by-skipping-duplicates
+    CollectionItem.insert_all!(newly_selected_collection_item_ids.map do |id|
+      max_order += 1
+      { resource_id: id,
+        resource_type: item_class.name,
+        collection_id: @collection.id,
+        order: max_order,
+        created_at: Time.zone.now,
+        updated_at: Time.zone.now }
+    end) # also no callbacks
+
+    # Now run a SOLR index on the added + removed entries
+    Sunspot.index! item_class.where(id: newly_selected_collection_item_ids + removed_collection_item_ids) if TeSS::Config.solr_enabled
+    true
+  end
+
+  def removed_collection_item_ids
+    reviewed_collection_item_ids - selected_collection_item_ids
+  end
+
+  def selected_collection_item_ids
+    Set.new(params[:item_ids]&.map(&:to_i))
+  end
+
+  def reviewed_collection_item_ids
+    Set.new(params[:reviewed_item_ids]&.map(&:to_i))
+  end
+
+  def newly_selected_collection_item_ids
+    selected_collection_item_ids - \
+      Set.new(CollectionItem.where(collection_id: @collection.id,
+                                   resource_id: selected_collection_item_ids,
+                                   resource_type: item_class.name).pluck(:resource_id))
```

Comment:
I still think this could simplified.

What about:
```ruby
method = "#{item_class.model_name.singular}_ids"
remove = params[:reviewed_item_ids]&.map(&:to_i)
keep_or_add = params[:item_ids]&.map(&:to_i)
@collection.send("#{method}=", (@collection.send(method) - remove) |  keep_or_add))
```

That should set the order, perform indexing etc. without having to do it manually

Reply:
I have written a new version, which uses subqueries to only select the items which did not exist yet, or which already existed (for add / rm respectively), using a query like:

```sql
SELECT "events".*
FROM "events"
WHERE "events"."id" = $1
  AND "events"."id" != $2
  AND (EXISTS
         (SELECT 1
          FROM "events"
          INNER JOIN "collection_items" ON "events"."id" = "collection_items"."resource_id"
          WHERE "collection_items"."collection_id" = 1
            AND "collection_items"."resource_type" = 'Event'
          ORDER BY "collection_items"."order" ASC))
```

Reply:
> Using item_ids instantiates all of the models (even the unchanged ones), which may be problematic for large collections. See collection_singular_ids= on https://apidock.com/rails/ActiveRecord/Associations/ClassMethods/has_many

Interesting! I would not have expected that.